pkg/proxy/kafka.go

```diff
@@ -0,0 +1,253 @@
+	source        ProxySource
+	listenPort    uint16
+	noMarker      bool
+	lookupNewDest destLookupFunc
+}
+
+// createKafkaRedirect creates a redirect with corresponding proxy
+// configuration. This will launch a proxy instance.
+func createKafkaRedirect(conf kafkaConfiguration) (Redirect, error) {
+	redir := &kafkaRedirect{
+		conf:    conf,
+		epID:    conf.source.GetID(),
+		ingress: conf.policy.Ingress,
+		nodeInfo: accesslog.NodeAddressInfo{
+			IPv4: nodeaddress.GetExternalIPv4().String(),
+			IPv6: nodeaddress.GetIPv6().String(),
+		},
+	}
+
+	if redir.conf.lookupNewDest == nil {
+		redir.conf.lookupNewDest = lookupNewDest
+	}
+
+	if err := redir.UpdateRules(conf.policy); err != nil {
+		return nil, err
+	}
+
+	marker := 0
+	if !conf.noMarker {
+		marker = GetMagicMark(redir.ingress)
+
+		// As ingress proxy, all replies to incoming requests must have the
+		// identity of the endpoint we are proxying for
+		if redir.ingress {
+			marker |= int(conf.source.GetIdentity())
+		}
+	}
+
+	// Listen needs to be in the synchronous part of this function to ensure that
+	// the proxy port is never refusing connections.
+	socket, err := listenSocket(fmt.Sprintf(":%d", redir.conf.listenPort), marker)
+	if err != nil {
+		return nil, err
+	}
+
+	redir.socket = socket
+
+	go func() {
+		for {
+			pair, err := socket.Accept()
+			select {
+			case <-socket.closing:
+				// Don't report errors while the socket is being closed
+				return
+			default:
+				if err != nil {
+					log.WithFields(log.Fields{
+						"listenPort": redir.conf.listenPort,
+					}).WithError(err).Error("Unable to accept connection")
+					continue
+				}
+			}
+
+			go redir.handleConnection(pair)
+		}
+	}()
+
+	return redir, nil
+}
+
+func (k *kafkaRedirect) canAccess(req *kafka.RequestMessage, srcIdentity uint32) bool {
+	rules := k.rules.GetRelevantRules(k.ingress, uint64(srcIdentity))
+	b, err := json.Marshal(rules.Kafka)
+	out := ""
+	if err != nil {
+		out = err.Error()
+	} else {
+		out = string(b)
+	}
+	log.WithFields(log.Fields{
+		"request": req.String(),
+	}).Debugf("Applying rules %s", out)
+	return req.MatchesRule(rules.Kafka)
+}
+
+func (k *kafkaRedirect) handleRequest(pair *connectionPair, req *kafka.RequestMessage) {
+	log.WithFields(log.Fields{
+		"id":      pair.String(),
+		"request": req.String(),
+	}).Debug("Handling Kafka request")
+
+	addr := pair.rx.conn.RemoteAddr()
+	if addr == nil {
+		log.Warning("RemoteAddr() is nil")
+		return
+	}
+
+	// retrieve identity of source together with original destination IP
+	// and destination port
+	srcIdentity, dstIPPort, err := k.conf.lookupNewDest(addr.String(), k.conf.listenPort)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"id":     pair.String(),
+			"source": addr.String(),
+		}).WithError(err).Error("Unable lookup original destination")
+		return
+	}
+
+	if !k.canAccess(req, srcIdentity) {
+		log.WithFields(log.Fields{
+			"id": pair.String(),
+		}).Debug("Kafka request is denied by policy")
```

Comment:
Should this be in `debug`?

Reply:
Imagine 100K Kafka requests/s going through and we info print this every time :-)